biapol_utilities/label/_matching_algorithms.py

```diff
@@ -0,0 +1,259 @@
+# -*- coding: utf-8 -*-
+
+import numpy as np
+from biapol_utilities import utilities
+
+
+def max_similarity(label_image_x, label_image_y, similarity_matrix):
+    """
+    Maximum-similarity algorithm for label-matching.
+
+    Matches labels in two input label images (label_image_x and label_image_y)
+    based on the maximal value in the similarity_matrix.
+
+    Parameters
+    ----------
+    label_image_x : ND-array, int
+        labelled input image which serves as reference image
+    label_image_y : ND-array, int
+        Labbelled input image, the labels of which will be matched up with the
+        labels in label_image_x based on the chosen input metric
+    similarity_matrix : 2D-array, float
+        matrix containing the metric that defines the similarity of labels
+        in label_image_x and label_image_y, with axis=0 referring to the labels
+        of label_image_x. In other words, the entry similarity_matrix[i, j]
+        corresponds to the "similarity" between label i in label_image_x and
+        label j in label_image_y.
+
+    Returns
+    -------
+    ND-array
+        Relabelled version of input_image_y. Unmatched labels are appended to
+        the set of matched labels and thus, the total number of labels may
+        increase.
+
+    """
+    # Remove background-label from similarity matrix
+    similarity_matrix = similarity_matrix[1:, 1:]
+
+    # Suppress non-maximal entries
+    similarity_matrix[similarity_matrix < similarity_matrix.max(axis=0)] = 0.0
+
+    mmax = label_image_x.max()
+
+    if similarity_matrix.size > 0:
+        # Pick value with highest IoU value
+        istitch = similarity_matrix.argmax(axis=1) + 1
+
+        # Find unpaired labels
+        ino = np.nonzero(similarity_matrix.max(axis=1) == 0.0)[0]
+
+        # append unmatched labels and background to lookup table
+        istitch[ino] = np.arange(mmax+1, mmax+len(ino)+1, 1, int)
+        mmax += len(ino)
+        istitch = np.append(np.array(0), istitch)
+
+        return istitch[label_image_y]
+
+    else:
+        raise ValueError('Similarity matrix dimension was expected to have '
+                         'size MxM but was found to have size '
+                         f'{similarity_matrix.size}. Check calculation of'
+                         'similarity matrix for errors.')
+
+
+def gale_shapley(label_image_x, label_image_y, similarity_matrix):
+    """
+    Gale-Shapley stable-marriage algorithm for label-matching.
+
+    Implementation of Gale-Shapley's solution of the stable-marriage problem
+    [0, 1] to match labels from two label images (`label_image_x` and
+    `label_image_y`) based on a mutual set of preferences that are passed as
+    `similarity_matrix`. Unmatchable labels in `label_image_x` maintain their
+    previous labels whereas unmatchable labels in `label_image_y` may be
+    assigned to new labels, thus potentially increasing the total number of
+    labels.
```

Comment:
Wow! Cool new stuff! Did you write all the code yourself? What speaks against reusing other implementatiosns such as 
* https://github.com/Vishal-Kancharla/Gale-Shapley-Algorithm/blob/master/StableMariages.py
* https://gist.github.com/joyrexus/9967709
* https://github.com/kartikeyas00/RAProblem/blob/master/Gale-Shap%20Sensitivity%20Analysis.py#L87

Reply:
Yeah... I did a bit of digging through some repositories, but wasn't really happy with aspects of them. Writing the core of the algorithm took me about half a day. The actual algorithm is just the part inside this while-loop:
```python
# Iterate over women and let them propose to men
while len(women) > 0:
```
It was tougher for me to get the "preference" vectors from the similarity matrix, and keeping track of indeces/labels/etc throughout the matching, which makes up for most of the code - I think I would have had to do some re-formatting anyway to fit other implementations. Plus, in the case of our problem, there are usually a bunch of "unmatchable" labels (i.e., labels with overlap too small to be matched or no overlaps at all) which need to be tracked.
Lastly, in comparison to some repos, I found it desirable to implement the algorithm as a function definition without the need to instantiate a class for it first.